### PR TITLE
MsgRequestPreVote messages from new splited regions shouldn't be dropped.

### PR DESCRIPTION
**Pull request by hicqu**
## What have you changed? (mandatory)
`MsgRequestPreVote`s are dropped incorrectly, so sometimes TiKV clusters will spend 10s to elect a leader for a splited region. This PR fix this, and remove some stale comments.

## What are the type of the changes? (mandatory)
Improvement.

## How has this PR been tested? (mandatory)
make dev.

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)
No.

## Does this PR affect tidb-ansible update? (mandatory)
No.

### CODE REVIEW

**Comment by siddontang** [comment]
Great catch!!!

**Comment by siddontang** [comment]
PTAL @Hoverbear

File: tests/failpoints_cases/test_split_region.rs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        Arc::from(AtomicBool::new(true)),
+    ));
+    cluster.sim.wl().add_send_filter(2, notifier);
+
+    fail::cfg("raftstore_follower_slow_split", "pause").unwrap();
+    cluster.must_split(&region, b"k2");
+    assert!(rx.recv_timeout(Duration::from_millis(100)).is_err());
+
+    fail::cfg("raftstore_follower_slow_split", "off").unwrap();
+    assert!(rx.recv_timeout(Duration::from_secs(5)).is_ok());
```
[the remark this answers is not in the document]
Comment (hicqu) [reply tests/failpoints_cases/test_split_region.rs]: yes, 5 secs is too long in test.

File: src/raftstore/store/fsm/peer.rs [hunk 2]
```diff
@@ -1035,7 +1029,21 @@ impl<T: Transport, C: PdClient> Store<T, C> {
             }
 
             new_peer.register_delegates();
-            self.region_peers.insert(new_region_id, new_peer);
+
+            if let Some(exist_peer) = self.region_peers.insert(new_region_id, new_peer) {
+                // Suppose a new node is added by conf change and the snapshot comes slowly.
+                // Then, the region splits and the first vote message comes to the new node
+                // before the old snapshot, which will create an uninitialized peer on the
+                // store. After that, the old snapshot comes, followed with a split proposal.
```
Comment (siddontang) [inline src/raftstore/store/fsm/peer.rs]: > followed with a split proposal

why another split proposal?

Comment (hicqu) [reply src/raftstore/store/fsm/peer.rs]: It's just the last split proposal, not a new split. Seems `followed with the last split proposal` is better.

**Comment by hicqu** [comment]
@BusyJay @zhangjinpeng1987 PTAL thanks.

File: src/raftstore/store/fsm/peer.rs [hunk 3]
```diff
@@ -1035,7 +1029,21 @@ impl<T: Transport, C: PdClient> Store<T, C> {
             }
 
             new_peer.register_delegates();
-            self.region_peers.insert(new_region_id, new_peer);
+
+            if let Some(exist_peer) = self.region_peers.insert(new_region_id, new_peer) {
```
Comment (BusyJay) [inline src/raftstore/store/fsm/peer.rs]: Prefer to keep it to the old place to check before bad things happen.

File: src/raftstore/store/fsm/peer.rs [hunk 4]
```diff
@@ -967,48 +967,39 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
             }
         };
+
         if is_leader {
             // Notify pd immediately to let it update the region meta.
             if let Err(e) = report_split_pd(&regions, &self.pd_worker) {
                 error!("{} failed to notify pd: {}", self.tag, e);
             }
+        } else {
+            fail_point!("raftstore_follower_slow_split");
```
Comment (BusyJay) [inline src/raftstore/store/fsm/peer.rs]: I think the failpoint should be set to apply worker.

**Comment by BusyJay** [comment]
There is still another bug. Consider following case: A cluster has only three stores and two regions A ["", "x"), B["x", ""). B has three peers (a, 1) , (b, 2) and (c, 3). If b is isolated, c is not created yet and store 3 is isolated before a campaigns. The quorum of B can never be formed.

File: src/raftstore/store/util.rs [hunk 5]
```diff
@@ -107,10 +107,42 @@ pub fn check_key_in_region(key: &[u8], region: &metapb::Region) -> Result<()> {
     }
 }
 
+/// `is_first_vote_msg` checks `msg` is the first vote (or prevote) message or not. It's used for
+/// when the message is receiving but there is no such region in `Store::region_peers` and the
```
Comment (BusyJay) [inline src/raftstore/store/util.rs]: s/receiving/received/

File: src/raftstore/store/util.rs [hunk 6]
```diff
@@ -107,10 +107,42 @@ pub fn check_key_in_region(key: &[u8], region: &metapb::Region) -> Result<()> {
...
+// There could be two cases:
+// 1. Target peer already exists but has not established communication with leader yet
+// 2. Target peer is added newly due to member change or region split, but it's not
+//    created yet
+// For both cases the region start key and end key are attached in RequestVote and
+// Heartbeat message for the store of that peer to check whether to create a new peer
+// when receiving these messages, or just to wait for a pending region split to perform
+// later.
+#[inline]
+pub fn msg_can_create_peer(msg: &eraftpb::Message) -> bool {
```
Comment (BusyJay) [inline src/raftstore/store/util.rs]: Any tests to cover it?

File: src/raftstore/store/peer.rs [hunk 7]
```diff
@@ -1886,16 +1886,11 @@ impl Peer {
         // Heartbeat message for the store of that peer to check whether to create a new peer
         // when receiving these messages, or just to wait for a pending region split to perform
         // later.
-        if self.get_store().is_initialized()
-            && (msg_type == MessageType::MsgRequestVote ||
-            // the peer has not been known to this leader, it may exist or not.
-            (msg_type == MessageType::MsgHeartbeat && msg.get_commit() == INVALID_INDEX))
-        {
+        if self.get_store().is_initialized() && msg_can_create_peer(&msg) {
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: Any tests to cover it?

File: tests/failpoints_cases/test_split_region.rs [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+    // After split, pre-vote message should be sent to peer 2, but we can't receive
+    // pre-vote response because it's buffered in pending_votes.
+    fail::cfg("apply_before_split_1_3", "pause").unwrap();
+    cluster.must_split(&region, b"k2");
+    assert!(rx1.recv_timeout(Duration::from_millis(100)).is_ok());
+    assert!(rx2.recv_timeout(Duration::from_millis(200)).is_err());
+
+    fail::cfg("apply_before_split_1_3", "off").unwrap();
+    assert!(rx2.recv_timeout(Duration::from_secs(2)).is_ok());
```
Comment (BusyJay) [inline tests/failpoints_cases/test_split_region.rs]: 2 seconds are too long.

File: src/raftstore/store/util.rs (hunk 6 above)
Comment (BusyJay) [reply src/raftstore/store/util.rs]: How about `is_initial_message`? These messages are the messages that a leader/candidate will send to a unknown peer. Message can't create a peer by themselves, only store can do it.

File: tests/failpoints_cases/test_split_region.rs [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+
+    // After the follower split success, it will response to the pending vote.
+    fail::cfg("apply_before_split_1_3", "off").unwrap();
+    assert!(rx.recv_timeout(Duration::from_millis(100)).is_ok());
+}
+
+// Filter prevote message and record the range.
+struct PrevoteRangeFilter {
+    filter: RegionPacketFilter,
+    start_key: Arc<Mutex<Vec<u8>>>,
```
Comment (BusyJay) [inline tests/failpoints_cases/test_split_region.rs]: Why not use a channel?

File: src/raftstore/store/util.rs [hunk 10]
```diff
@@ -1143,10 +1179,25 @@ mod tests {
...
+        }
+    }
+
+    #[test]
+    fn test_msg_can_create_peer() {
+        let tbl = vec![
+            (MessageType::MsgRequestVote, INVALID_INDEX, true),
+            (MessageType::MsgRequestPreVote, INVALID_INDEX, true),
+            (MessageType::MsgHeartbeat, INVALID_INDEX, false),
+            (MessageType::MsgHeartbeat, 100, true),
```
Comment (BusyJay) [inline src/raftstore/store/util.rs]: I think it should be false.

File: src/raftstore/store/util.rs [hunk 11]
```diff
@@ -1127,6 +1158,11 @@ mod tests {
                 peer_storage::RAFT_INIT_LOG_TERM + 1,
                 true,
             ),
+            (
+                MessageType::MsgRequestPreVote,
+                peer_storage::RAFT_INIT_LOG_TERM + 1,
```
Comment (BusyJay) [inline src/raftstore/store/util.rs]: What if term is `RAFT_INIT_LOG_TERM`?

Comment (hicqu) [reply src/raftstore/store/util.rs]: Fixed

**Pull request merged**